py/gc.c

```diff
@@ -235,6 +235,106 @@ void gc_add(void *start, void *end) {
     // Add this area to the linked list
     prev_area->next = area;
 }
+
+#if MICROPY_GC_SPLIT_HEAP_AUTO
+// Try to automatically add a heap area large enough to
+// fulfill 'failed_alloc'
+STATIC bool gc_try_add_heap(size_t failed_alloc) {
+    // Equation to find a heap large enough to hold failed_alloc is derived from
+    // the equation for gc_setup_area(), above:
+    //
+    // T = A + F + P
+    // P == failed_alloc rounded up to next block size
+    // A = P / (BLOCKS_PER_ATB * BYTES_PER_BLOCK
+    // F = A * BLOCKS_PER_ATB / BLOCKS_PER_FTB
+    size_t p_blocks = (failed_alloc + BYTES_PER_BLOCK - 1) / BYTES_PER_BLOCK;
+    // Round up to use full allocation block(s)
+    p_blocks = (p_blocks + BLOCKS_PER_ATB - 1) & ~(BLOCKS_PER_ATB - 1);
+
+    // Unclear why these extra blocks needed, but necessary for the new heap
+    // pool to be large enough
+    p_blocks += BLOCKS_PER_ATB;
+
+    size_t p = p_blocks * BYTES_PER_BLOCK; // should match gc_pool_end - gc_pool_start
+
+    size_t a = p_blocks / BLOCKS_PER_ATB;
+
+    #if MICROPY_ENABLE_FINALISER
+    // should match gc_finaliser_table_byte_len
+    size_t f = (p_blocks + BLOCKS_PER_FTB - 1) / BLOCKS_PER_FTB;
+    #else
+    size_t f = 0;
+    #endif
+
+    // T = ... (plus implementation overheads)
+    size_t needed = p + a + f + sizeof(mp_state_mem_area_t) + ALLOC_TABLE_GAP_BYTE;
```

Comment:
Rather than the complexity (and code size) of calculating all of this, how about just doing:

`size_t needed = MAX(1024, failed_alloc * 3 / 2)`

The idea being:
 - If the allocation is more than 1kiB, then 1.5x is guaranteed to be enough for the ATB+FTB etc, bypassing the complicated calculation.
 - If the allocation is less than 1kiB, then it's very likely that we're going to do a bigger allocation anyway with the "doubling the heap" logic below.

So this means that if the underlying system heap (e.g. IDF) has less than 1kiB free, we cannot satisfy any further allocations, even if in theory it were possible. But we don't really want the IDF to get into this state anyway, and I think these allocations _should_ fail.

Reply:
I've written those bullet points poorly... sorry. But I think the logic is sound.

The point is that above a certain amount (which is way less than 1024 bytes) 1.5x is always going to be enough.

Reply:
> Under the proposed scheme, you can't allocate such a buffer (tries to allocate 4.5MB, fails).

No, it would still work under the proposed scheme: it would just cap the allocation at the maximum available (like it currently does).

The only downside I see of this simpler approach is that this case (allocating 3MB when there's 4MB total available) doesn't leave any for the IDF (or other system memory on another system).  But is that an important case?  You could just allocate IDF memory first (eg SSLContext buffers).

Reply:
> There's no requirement that we use 2x as the scale factor though. I think 1.05 is safe for all allocations above 1kiB for example.

You mean "no requirement that use 1.5x as the scale factor" here yeah, @jimmo? I agree that's a reasonable simplification (assume 5% heap metadata overhead), compared to trying to mirror the actual `gc_setup` logic here.

(I did initially try to find a way to unify the two sets of heap layout calculations into one function - but as you're probably very aware it's hard to combine them as `gc_setup` is solving for different unknown variables in the equation.)

Reply:
Right, I see, the hard problem is how to know when you have just enough (or just not enough) system memory to satisfy the current request.  So the application needs X bytes and the system has X+epsilon (where epsilon is small) bytes available.  What is the value of epsilon (which is a function of X) for which the allocation should succeed/fail.

Indeed we need to compute this, and it would be good if it's a simple calculation.  It can be conservative (epsilon larger than necessary) but not too conservative that it unnecessarily fails.  As @jimmo says, probably `X=MAX(1024, X)` and `epsilon=0.05*X` will be OK.

Reply:
@dpgeorge @jimmo PTAL again. I've gone with 103%, with a comment inline to explain the logic.

Reply:
Oh sorry I only just saw your latest response @dpgeorge , Github seems to not always auto-load these on an open page for me.

I think 0.03125% (1/32) should be OK though - ATBs are 1/64 of total size, and FTBs are 1/128 of total size, both rounding up to the next block increment. Plus a few dozen bytes for the area struct and the extra separator byte.

Have implemented the "max 1024" in a slightly different way by bailing out if there's not 1024 bytes of free system heap. In other cases where `failed_to_alloc` is small, the second part of the function will pick a much larger new heap are anyhow.

Reply:
I'm probably overthinking it now, but if the logic in previous comments seems confusing then a better way to express this might be:

```c
size_t needed = failed_alloc + MIN(512, failed_alloc/32);
```

This keeps the meaning of needed coupled to "at least how many bytes of heap do we need to fulfill this allocation?", while accounting for enough overhead on smalled `failed_alloc` values *and* keeping the code size down.

(That said, I think in practical terms all these versions will behave much the same.)

Reply:
Updated with the refined version of the above that we talked about on Discord.